vllm/interfaces/communicator.py

```diff
@@ -0,0 +1,69 @@
+# communicator interface, as proposed in
+# https://github.com/vllm-project/vllm/issues/3587
+# `Communicator` is responsible for communicating **large tensor data**
+# between multiple devices. This functionality is usually provided by
+# vendors, e.g. NCCL from NVIDIA, RCCL from AMD.
+# Put it simple, this is for data-plane communication.
+
+from typing import Any, Optional
+
+import torch
+from torch.distributed import ReduceOp
+
+from vllm.interfaces.coordinator import Coordinator
+
+
+class Communicator(object):
```

Comment:
why does this extend `object`?

Reply:
> I feel we should avoid super().method() for user implementations as much as possible, as then the code relevant to the layer is completely contained within the implementation and not spread between implementation and interface.

Technically it is doable. But then some code will be replicated over many implementations. There is a trade-off here.

> Is it possible with this design to run the Engine in a different process?

Yes we can. The launcher design is general, and we can have both CPU workers and GPU workers. Although we need to think about how they coordinate with each other.

> why does this extend object?
see https://stackoverflow.com/questions/4015417/why-do-python-classes-inherit-object

Reply:
👍 

btw python 2 is EOL'd a long time ago (in 2020 IIRC), we don't need to extend object anymore!

Reply:
I disagree that code will be replicated if not using `super().method()`. There is a third way--factor out common utilities that can be used in multiple implementations. Basically, composition vs. inheritance.